src/include/common/performance_counter.h

```diff
@@ -0,0 +1,188 @@
+ */
+#define PC_HELPER_DEFINE_INCREMENT(MemberType, MemberName) \
+  void Inc_##MemberName() { ++MemberName; }
+
+/**
+ * This macro defines a Dec_MemberName() function which decrements MemberName.
+ */
+#define PC_HELPER_DEFINE_DECREMENT(MemberType, MemberName) \
+  void Dec_##MemberName() { --MemberName; }
+
+/**
+ * This macro defines a Get_MemberName() function which returns MemberName.
+ * If performance counters are disabled, it always returns 0.
+ */
+#define PC_HELPER_DEFINE_GET(MemberType, MemberName) \
+  MemberType Get_##MemberName() { return MemberName.load(); }
+
+/*
+ * Performance counter functions.
+ * These macros add new functionality.
+ * Please maintain alphabetical A-Z order.
+ */
+/**
+ * This macro loads ClassName.MemberName from the JSON object.
+ *
+ * Assumed in scope:
+ *  json &j
+ */
+#define PC_FN_JSON_FROM(MemberType, MemberName) MemberName.store(j.at("Counters").at(#MemberName).get<MemberType>());
+
+/**
+ * This macro writes ClassName.MemberName into the JSON object.
+ *
+ * Assumed in scope:
+ *  json output
+ */
+#define PC_FN_JSON_TO(MemberType, MemberName) output["Counters"][#MemberName] = MemberName.load();
+
+/**
+ * This macro zeroes out ClassName.MemberName.
+ */
+#define PC_FN_ZERO(MemberType, MemberName) MemberName.store(0);
+#endif  // NDEBUG
+
+/*
+ * PerformanceCounter implementation details.
+ *
+ * We rely on the fact that macros can call macros as arguments.
+ * More concretely, given the following macro definitions:
+ *
+ * #define NETWORK_MEMBERS(f) \
+ *    f(uint64_t, requests_received) \
+ *    f(uint32_t, connections_opened)
+ *
+ * #define MAKE_MEMBER(MemberType, MemberName) \
+ *    std::atomic<MemberType> MemberName{0};
+ *
+ * The preprocessor will expand NETWORK_MEMBERS(MAKE_MEMBER) to
+ *    std::atomic<uint64_t> requests_received{0};
+ *    std::atomic<uint32_t> connections_opened{0};
+ *
+ * Furthermore, you have access to the surrounding scope's variables,
+ * though you have to be careful with any assumptions you make there.
+ */
+
+/*
+ * This macro creates a PerformanceCounter in its current namespace as the class ClassName.
+ *
+ * The actual counters are defined by passing in MemberList, a macro of the form
+ *      #define MEMBER_LIST(f) f(type1, name1) f(type2, name2) ... f(typeN, nameN)
+ * Note that all of the types should be integral types. An example:
+ *      #define NETWORK_MEMBERS(f) f(uint64_t, requests_received) f(uint32_t, connections_opened)
+ *
+ * You can use the resulting PerformanceCounter as a class in its own right. For example,
+ *      #define MAKE_PERFORMANCE_COUNTER(NetworkCounter, NETWORK_MEMBERS)
+ * will make the following code valid:
+ *      NetworkCounter nc;
+ *      nc.requests_received++;
+ *      nc.connections_opened.store(100);
+ *
+ * Note that every class member is wrapped in std::atomic.
+ */
+#define MAKE_PERFORMANCE_COUNTER(ClassName, MemberList)                               \
```

Comment:
Name could use adjustment.   MAKE should probably be DEFINE (as in declaration vs. definition)
COUNTER should be COUNTERS, since the argument is a list.